src/Chainweb/Pact/PactService/ExecBlock.hs

```diff
@@ -376,48 +378,68 @@ applyPactCmds
     -> Vector ChainwebTransaction
     -> Miner
     -> ModuleCache
+    -> Maybe P.Gas
     -> PactServiceM cas (Vector (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value])))
-applyPactCmds isGenesis env cmds miner mc =
-    V.fromList . toList . sfst <$> V.foldM f (T2 mempty mc) cmds
-  where
-    f  (T2 dl mcache) cmd = applyPactCmd isGenesis env cmd miner mcache dl
+applyPactCmds isGenesis env cmds miner mc blockGas = do
+    evalStateT (V.mapM (applyPactCmd isGenesis env miner) cmds) (T2 mc blockGas)
 
--- | Apply a single Pact command
 applyPactCmd
-    :: Bool
-    -> P.PactDbEnv p
-    -> ChainwebTransaction
-    -> Miner
-    -> ModuleCache
-    -> DList (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))
-    -> PactServiceM cas
-       (T2 (DList (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))) ModuleCache)
-applyPactCmd isGenesis dbEnv cmdIn miner mcache dl = do
-    logger <- view psLogger
-    gasModel <- view psGasModel
-    v <- view psVersion
-
-    handle onBuyGasFailure $ do
-      T2 result mcache' <- if isGenesis
-        then liftIO $! applyGenesisCmd logger dbEnv P.noSPVSupport (payloadObj <$> cmdIn)
-        else do
-          pd <- getTxContext (publicMetaOf $ payloadObj <$> cmdIn)
-          spv <- use psSpvSupport
-          liftIO $! applyCmd v logger dbEnv miner (gasModel pd) pd spv cmdIn mcache
-
-      when isGenesis $
-        updateInitCache mcache'
+  :: Bool
+  -> P.PactDbEnv p
+  -> Miner
+  -> ChainwebTransaction
+  -> StateT 
+      (T2 ModuleCache (Maybe P.Gas)) 
+      (PactServiceM cas) 
+      (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))
+applyPactCmd isGenesis env miner cmd = StateT $ \(T2 mcache blockGasRemaining) -> do
+  logger <- view psLogger
+  gasModel <- view psGasModel
+  v <- view psVersion
+  let
+    onBuyGasFailure (BuyGasFailure f) = pure $! (Left f, T2 mcache blockGasRemaining)
+    onBuyGasFailure e = throwM e
+    requestedTxGasLimit = view cmdGasLimit (payloadObj <$> cmd)
+    -- notice that we add 1 to the remaining block gas here, to distinguish
+    -- the cases "tx used exactly as much gas remained in the block" and "tx
+    -- attempted to use more gas than remains in the block" while still not
+    -- letting it use significantly more gas than remains in the block.  for
+    -- example: tx has a tx gas limit of 10000. the block has 5000 remaining
+    -- gas.  the tx is applied with a tx gas limit of 5001. if it uses 5001,
+    -- that's illegal; if it uses 5000 or less, that's legal.
+    newTxGasLimit = case blockGasRemaining of
+      Nothing -> requestedTxGasLimit
+      Just blockGasLimit -> min (fromIntegral (succ blockGasLimit)) requestedTxGasLimit
```

Comment:
`blockGasRemaining'` or something, had a bit of confusion here, but otherwise :+1: